### PR TITLE
Api request helper refactor

### PR DESCRIPTION
**Pull request by supermassive**
Resolves https://github.com/brave/brave-browser/issues/24582

Adding `base::Value value_body()` to `APIRequestResult` and removing json parsing from callers. Might be reviewed by commits.

## Submitter Checklist:

- [x] I confirm that no security/privacy review [is needed](https://github.com/brave/brave-browser/wiki/Security-reviews), or that I have [requested](https://github.com/brave/security/issues/new/choose) one
- [x] There is a [ticket](https://github.com/brave/brave-browser/issues) for my issue
- [x] Used Github [auto-closing keywords](https://docs.github.com/en/github/managing-your-work-on-github/linking-a-pull-request-to-an-issue) in the PR description above
- [x] Wrote a good [PR/commit description](https://google.github.io/eng-practices/review/developer/cl-descriptions.html)
- [x] Squashed any review feedback or "fixup" commits before merge, so that history is a record of what happened in the repo, not your PR
- [x] Added appropriate labels (`QA/Yes` or `QA/No`; `release-notes/include` or `release-notes/exclude`; `OS/...`) to the associated issue
- [x] Checked the PR locally: `npm run test -- brave_browser_tests`, `npm run test -- brave_unit_tests`, `npm run lint`, `npm run gn_check`, `npm run tslint`
- [ ] Ran `git rebase master` (if needed)

## Reviewer Checklist:

- [ ] A security review [is not needed](https://github.com/brave/brave-browser/wiki/Security-reviews), or a link to one is included in the PR description
- [ ] New files have MPL-2.0 license header
- [ ] Adequate test coverage exists to prevent regressions
- [ ] Major classes, functions and non-trivial code blocks are well-commented
- [ ] Changes in component dependencies are properly reflected in `gn`
- [ ] Code follows the [style guide](https://chromium.googlesource.com/chromium/src/+/HEAD/styleguide/c++/c++.md)
- [ ] Test plan is specified in PR before merging

## After-merge Checklist:

- [ ] The associated issue milestone is set to the smallest version that the
  changes has landed on
- [ ] All relevant documentation has been updated, for instance:
  - [ ] https://github.com/brave/brave-browser/wiki/Deviations-from-Chromium-(features-we-disable-or-remove)
  - [ ] https://github.com/brave/brave-browser/wiki/Proxy-redirected-URLs
  - [ ] https://github.com/brave/brave-browser/wiki/Fingerprinting-Protections
  - [ ] https://github.com/brave/brave-browser/wiki/Brave%E2%80%99s-Use-of-Referral-Codes
  - [ ] https://github.com/brave/brave-browser/wiki/Custom-Headers
  - [ ] https://github.com/brave/brave-browser/wiki/Web-Compatibility-Exceptions-in-Brave
  - [ ] https://github.com/brave/brave-browser/wiki/QA-Guide
  - [ ] https://github.com/brave/brave-browser/wiki/P3A

### CODE REVIEW

File: components/api_request_helper/api_request_helper.cc [hunk 1]
```diff
@@ -202,25 +227,26 @@ void APIRequestHelper::OnResponse(
...
+                                             final_url));
     return;
   }
   auto& raw_body = *response_body;
   if (conversion_callback) {
     auto converted_body = std::move(conversion_callback).Run(raw_body);
     if (!converted_body) {
       std::move(callback).Run(APIRequestResult(
-          422, std::move(raw_body), std::move(headers), error_code, final_url));
+          422, "", base::Value(), std::move(headers), error_code, final_url));
```
Comment (supermassive) [inline components/api_request_helper/api_request_helper.cc:240]: Non-sanitized raw_body could potentially leak to caller. Now just return empty body.

File: components/api_request_helper/api_request_helper_unittest.cc [hunk 2]
```diff
@@ -101,11 +115,7 @@ class ApiRequestHelperUnitTest : public testing::Test {
 
 TEST_F(ApiRequestHelperUnitTest, SanitizedRequest) {
   std::string expected_sanitized_response =
-#if BUILDFLAG(IS_ANDROID)
-      "{\"id\":1,\"jsonrpc\":\"2.0\",\"result\":1.8446744073709552E19}";
-#else
       "{\"id\":1,\"jsonrpc\":\"2.0\",\"result\":1.8446744073709552e+19}";
-#endif
```
Comment (supermassive) [inline components/api_request_helper/api_request_helper_unittest.cc:108]: Now string body is always a result of JSONWriter::Write, so is platform indepenend.

File: components/brave_wallet/browser/asset_ratio_response_parser.cc [hunk 3]
```diff
@@ -177,30 +168,7 @@ bool ParseAssetPriceHistory(const std::string& json,
   return true;
 }
 
-std::string ParseEstimatedTime(const std::string& json) {
```
Comment (supermassive) [inline components/brave_wallet/browser/asset_ratio_response_parser.cc:180]: `ParseEstimatedTime` was never referenced

**Review by simonhong — APPROVED** [review]
++ 👍🏼

File: components/api_request_helper/api_request_helper.h [hunk 4]
```diff
@@ -30,32 +31,38 @@ class APIRequestResult {
   APIRequestResult();
   APIRequestResult(int response_code,
                    std::string body,
+                   base::Value value_body,
```
Comment (iefremov) [inline components/api_request_helper/api_request_helper.h:34]: can you please write a comment describing all the parameters, esp body/value_body and final_url

File: components/api_request_helper/api_request_helper.cc [hunk 5]
```diff
@@ -19,27 +20,40 @@ namespace api_request_helper {
...
-                const base::flat_map<std::string, std::string>& headers,
-                int error_code,
-                GURL final_url,
-                APIRequestHelper::ResultCallback result_callback,
-                data_decoder::JsonSanitizer::Result result) {
-  std::string response_body;
-  if (result.error) {
-    VLOG(1) << "Response validation error:" << *result.error;
+void OnParseJsonIsolated(
+    const int http_code,
```
Comment (iefremov) [inline components/api_request_helper/api_request_helper.cc:24]: redundant const

File: components/brave_wallet/browser/asset_ratio_response_parser_unittest.cc [hunk 6]
```diff
@@ -7,32 +7,40 @@
...
 #include "brave/components/brave_wallet/browser/asset_ratio_response_parser.h"
 #include "brave/components/brave_wallet/browser/brave_wallet_utils.h"
 #include "brave/components/brave_wallet/common/brave_wallet.mojom.h"
 #include "testing/gtest/include/gtest/gtest.h"
 
 namespace brave_wallet {
 
+namespace {
+base::Value ToValue(const std::string& json) {
+  return base::JSONReader::Read(json).value_or(base::Value());
```
Comment (iefremov) [inline components/brave_wallet/browser/asset_ratio_response_parser_unittest.cc:22]: perhaps some utils from `values_test_utils.h`could be handier, e.g. `ParseJson` or `ParseJsonDict`

**Review by iefremov — APPROVED** [review]

File: components/brave_wallet/browser/asset_ratio_response_parser_unittest.cc (hunk 6 above)
Comment (supermassive) [reply components/brave_wallet/browser/asset_ratio_response_parser_unittest.cc:22]: TIL values_test_util exists :)

**Pull request merged**